x-pack/metricbeat/module/azure/azure.go

```diff
@@ -63,3 +65,63 @@ func newModule(base mb.BaseModule) (mb.Module, error) {
 	}
 	return &base, nil
 }
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	mb.BaseMetricSet
+	Client *Client
+}
+
+// NewMetricSet will instantiate a new azure metricset
+func NewMetricSet(base mb.BaseMetricSet) (*MetricSet, error) {
+	metricsetName := base.Name()
+	cfgwarn.Beta("The azure %s metricset is beta.", metricsetName)
+	var config Config
+	err := base.Module().UnpackConfig(&config)
+	if err != nil {
+		return nil, errors.Wrap(err, "error unpack raw module config using UnpackConfig")
+	}
+
+	//validate config based on metricset
+	switch base.Name() {
+	case nativeMetricset:
+		// resources must be configured for the monitor metricset
+		if len(config.Resources) == 0 {
+			return nil, errors.Errorf("no resource options defined: module azure - %s metricset", metricsetName)
+		}
+	default:
+		// validate config resource options entered, no resource queries allowed for the compute_vm and compute_vm_scaleset metricsets
+		for _, resource := range config.Resources {
+			if resource.Query != "" {
+				return nil, errors.Errorf("error initializing the monitor client: module azure - %s metricset. No queries allowed, please select one of the allowed options", metricsetName)
+			}
+		}
+
+	}
+	// instantiate monitor client
+	monitorClient, err := NewClient(config)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error initializing the monitor client: module azure - %s metricset", metricsetName)
+	}
+	return &MetricSet{
+		BaseMetricSet: base,
+		Client:        monitorClient,
+	}, nil
+}
+
+// FetchValues methods implements the data gathering and data conversion to the right metricset
+func (m *MetricSet) FetchValues(fn mapMetric, report mb.ReporterV2) error {
+	err := m.Client.InitResources(fn, report)
```

Comment:
If `fn` would be an attribute of the metricset I think that this `FetchValues` could be the `Fetch` method of all derived metricsets.
```suggestion
	err := m.Client.InitResources(m.mapMetricFn, report)
```

Reply:
I like the suggestion, I have made the changes

Reply:
Thanks!